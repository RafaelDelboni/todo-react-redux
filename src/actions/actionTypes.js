// Action types constants
export const LOAD_TODOS_SUCCESS = 'LOAD_TODOS_SUCCESS';
export const CREATE_TODO_SUCCESS = 'CREATE_TODO_SUCCESS';
export const UPDATE_TODO_SUCCESS = 'UPDATE_TODO_SUCCESS';
export const DELETE_TODO_SUCCESS = 'DELETE_TODO_SUCCESS';

export const LOAD_TODOS_ERROR = 'LOAD_TODOS_ERROR';
export const CREATE_UPDATE_TODO_ERROR = 'CREATE_UPDATE_TODO_ERROR';
export const DELETE_TODO_ERROR = 'DELETE_TODO_ERROR';
