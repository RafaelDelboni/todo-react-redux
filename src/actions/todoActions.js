import * as types from './actionTypes';
import todoApi from '../api/todoApi.fake';
import {beginAjaxCall, ajaxCallError} from './ajaxStatusActions';

export function loadTodosSuccess(todos) {
  return {
    type: types.LOAD_TODOS_SUCCESS,
    todos
  };
}

export function updateTodoSuccess(todo) {
  return {
    type: types.UPDATE_TODO_SUCCESS,
    todo
  };
}

export function saveTodoSuccess(todo) {
  return {
    type: types.CREATE_TODO_SUCCESS,
    todo
  };
}

export function deleteTodoSuccess(todo) {
  return {
    type: types.DELETE_TODO_SUCCESS,
    todo
  };
}

export function loadTodosError(error) {
  return {
    type: types.LOAD_TODOS_ERROR,
    error
  };
}

export function saveTodoError(error) {
  return {
    type: types.CREATE_UPDATE_TODO_ERROR,
    error
  };
}

export function deleteTodoError(error) {
  return {
    type: types.DELETE_TODO_ERROR,
    error
  };
}

export function loadTodos() {
  return function (dispatch) {
    dispatch(beginAjaxCall());
    return todoApi.getAllTodos().then(todos => {
      dispatch(loadTodosSuccess(todos));
    }).catch(error => {
      dispatch(loadTodosError(error));
      dispatch(ajaxCallError());
    });
  };
}

export function saveTodo(todo) {
  return function(dispatch) {
    dispatch(beginAjaxCall());
    return todoApi.saveTodo(todo).then(savedTodo => {
      todo._id ? dispatch(updateTodoSuccess(savedTodo)) : dispatch(saveTodoSuccess(savedTodo));
      return savedTodo;
    }).catch(error => {
      dispatch(saveTodoError(error));
      dispatch(ajaxCallError());
    });
  };
}

export function deleteTodo(todo) {
  return function(dispatch) {
    dispatch(beginAjaxCall());
    return todoApi.deleteTodo(todo).then(deletedTodo => {
      dispatch(deleteTodoSuccess(deletedTodo));
      return deletedTodo;
    }).catch(error => {
      dispatch(deleteTodoError(error));
      dispatch(ajaxCallError());
    });
  };
}
