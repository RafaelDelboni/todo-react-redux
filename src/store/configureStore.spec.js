import {createStore} from 'redux';
import rootReducer from '../reducers';
import initialState from '../reducers/initialState';
import * as todoActions from '../actions/todoActions';

describe('Store', () => {
  it('should handle creating todos', () => {
    const store = createStore(rootReducer, initialState);
    const todo = {
      description: "First todo",
      done: false
    };

    const action = todoActions.saveTodoSuccess(todo);
    store.dispatch(action);

    const actual = store.getState().todos[0];
    const expected = {
      description: "First todo",
      done: false
    };

    expect(actual).toEqual(expected);
  });

  it('should handle updating todos', () => {
    const store = createStore(rootReducer, initialState);
    const todo = {
      _id: 2,
      description: "Second todo",
      done: false
    };

    const actionCreate = todoActions.saveTodoSuccess(todo);
    store.dispatch(actionCreate);

    const updatedTodo = {
      _id: 2,
      description: "Second todo updated",
      done: true
    };

    const actionUpdate = todoActions.updateTodoSuccess(updatedTodo);
    store.dispatch(actionUpdate);

    const actual = store.getState().todos[0];

    expect(actual).toMatchSnapshot();
  });
  
  it('should handle deleting todos', () => {
    const store = createStore(rootReducer, initialState);
    const todo = {
      _id: 3,
      description: "Third todo",
      done: true
    };

    const actionCreate = todoActions.saveTodoSuccess(todo);
    store.dispatch(actionCreate);

    const actualBefore = store.getState().todos[0];

    expect(actualBefore).toMatchSnapshot();

    const actionDelete = todoActions.deleteTodoSuccess(todo);
    store.dispatch(actionDelete);

    const actual = store.getState().todos.length;
    const expected = 0;

    expect(actual).toEqual(expected);
  });
});