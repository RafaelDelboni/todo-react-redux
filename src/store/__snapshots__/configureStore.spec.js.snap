exports[`Store should handle deleting todos 1`] = `
Object {
  "_id": 3,
  "description": "Third todo",
  "done": true,
}
`;

exports[`Store should handle updating todos 1`] = `
Object {
  "_id": 2,
  "description": "Second todo updated",
  "done": true,
}
`;
